src/rpc/core_rpc_server_commands_defs.h

```diff
@@ -2537,69 +2538,40 @@ namespace cryptonote
   {
     struct request_t
     {
-      uint64_t height; // The height to query the quorum state for.
+      std::vector<uint64_t> heights; // Array of heights to query the quorums for, omit the height to request the latest quorum
       BEGIN_KV_SERIALIZE_MAP()
-        KV_SERIALIZE(height)
-      END_KV_SERIALIZE_MAP()
-    };
-    typedef epee::misc_utils::struct_init<request_t> request;
-
-    struct response_t
-    {
-      std::string status;                     // Generic RPC error code. "OK" is the success value.
-      std::vector<std::string> quorum_nodes;  // Array of public keys identifying service nodes which are being tested for the queried height.
-      std::vector<std::string> nodes_to_test; // Array of public keys identifying service nodes which are responsible for voting on the queried height.
-      bool untrusted;                         // If the result is obtained using bootstrap mode, and therefore not trusted `true`, or otherwise `false`.
-
-      BEGIN_KV_SERIALIZE_MAP()
-        KV_SERIALIZE(status)
-        KV_SERIALIZE(quorum_nodes)
-        KV_SERIALIZE(nodes_to_test)
-        KV_SERIALIZE(untrusted)
-      END_KV_SERIALIZE_MAP()
-    };
-    typedef epee::misc_utils::struct_init<response_t> response;
-  };
-
-  LOKI_RPC_DOC_INTROSPECT
-  // Get the quorum state which is the list of public keys of the nodes 
-  // who are voting, and the list of public keys of the nodes who are being tested.
-  struct COMMAND_RPC_GET_QUORUM_STATE_BATCHED
-  {
-    struct request_t
-    {
-      uint64_t height_begin; // The starting height (inclusive) to query the quorum state for.
-      uint64_t height_end;   // The ending height (inclusive) to query the quorum state for.
-
-      BEGIN_KV_SERIALIZE_MAP()
-        KV_SERIALIZE(height_begin)
-        KV_SERIALIZE(height_end)
+        KV_SERIALIZE(heights)
       END_KV_SERIALIZE_MAP()
     };
     typedef epee::misc_utils::struct_init<request_t> request;
 
-    struct response_entry
+    struct quorums_for_height
     {
-      uint64_t height;                        // The height of this quorum state that was queried.
-      std::vector<std::string> quorum_nodes;  // Array of public keys identifying service nodes which are being tested for the queried height.
-      std::vector<std::string> nodes_to_test; // Array of public keys identifying service nodes which are responsible for voting on the queried height.
-
+      uint64_t height;                             // The height the quorums are relevant for
+      service_nodes::testing_quorum obligation;    // Quorum for checking Service Nodes have performed their duties and should not be voted off
+      service_nodes::testing_quorum checkpointing; // Quorum for Service Nodes responsible for participating in making checkpoints
       BEGIN_KV_SERIALIZE_MAP()
         KV_SERIALIZE(height)
-        KV_SERIALIZE(quorum_nodes)
-        KV_SERIALIZE(nodes_to_test)
+        KV_SERIALIZE(obligation)
+        KV_SERIALIZE(checkpointing)
```

Comment:
I don't like this interface change: the typical use case of a caller will be to get *either* an obligation *or* a checkpointing quorum, and so this will end up retrieving extra data and putting the data I care about one level deeper in the structure.

We should rather parameterize the quorum choice and return just that one quorum (but renamed to `validators` and `workers` instead of `quorum_nodes` and `nodes_to_test`).  In the case where you want both at once for the same height (which seems relatively rare given the significantly different purposes of the quorums) it is not a big deal to make two RPC calls.

Reply:
The actual only use case I have seen here for this is the block explorer and that wants all the relevant quorums for that height. I agree with making it customisable is nice but generally apprehensive about designing for future use cases that haven't been realised yet and it's not obvious how people are wanting to use this information either.

I'll make it parameterized since its does give more control, just preferred it was done at a later step when we actually have legitimate use cases outside of the 1 we use.